src/main/java/com/puppycrawl/tools/checkstyle/xpath/SuppressionsFileGenerator.java

```diff
@@ -0,0 +1,176 @@
+import com.puppycrawl.tools.checkstyle.api.LocalizedMessage;
+
+/**
+ * Creates <b>suppressions.xml</b> file, based on violations occurred.
+ *
+ * @author Timur Tibeyev.
+ */
+public class SuppressionsFileGenerator extends AutomaticBean implements AuditListener {
+
+    /**
+     * Helper writer that allows easy encoding and printing.
+     */
+    private final PrintWriter writer;
+
+    /** The distance between tab stop position. */
+    private final int tabWidth;
+
+    /** The delimiter between xpath queries. */
+    private static final String DELIMITER = " | \n";
+
+    private final String[] supportedChecks = {
+            "CyclomaticComplexity", "DeclarationOrder", "DefaultComesLast",
+            "ExplicitInitialization", "FallThrough", "HiddenField", "IllegalThrows",
+            "ImportControl", "JavadocVariable", "LeftCurly", "MethodParamPad",
+            "MultipleVariableDeclarations", "NestedForDepth", "NestedIfDepth", "NestedTryDepth",
+            "NPathComplexity", "OneStatementPerLine", "OuterTypeNumber", "RequireThis", "RightCurly"
+    };
+
+    /**
+     * Since currently not all checks support xpath suppressions, this method returns true if check
+     * supports xpath and false otherwise.
+     * @param checkName the name of the check.
+     * @return returns true if name is valid
+     */
+    private boolean isValidCheckName(String checkName) {
+        boolean result = false;
+        for (String supportedCheck : supportedChecks) {
+            if (checkName.contains(supportedCheck)) {
+                result = true;
+                break;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Creates a new {@code SuppressionFileGenerator} instance.
+     * Sets the output to a defined stream.
+     * @param outputStream the stream to write logs to.
+     * @param tabWidth distance between tab stop position
+     */
+    public SuppressionsFileGenerator(OutputStream outputStream,
+                                     int tabWidth) {
+        this.tabWidth = tabWidth;
+        writer = new PrintWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+
+        writer.println("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
+        writer.println("<!DOCTYPE suppressions PUBLIC");
+        writer.println("    \"-//Puppy Crawl//DTD Suppressions Xpath Experimental 1.2//EN\"");
+        writer.println("    \"http://checkstyle.sourceforge.net/dtds"
+                + "/suppressions_1_2_xpath_experimental.dtd\">");
+    }
+
+    @Override
+    public void auditStarted(AuditEvent event) {
+        writer.println("<suppressions>");
+    }
+
+    @Override
+    public void auditFinished(AuditEvent event) {
+        writer.println("</suppressions>");
+        writer.flush();
+    }
+
+    @Override
+    public void fileStarted(AuditEvent event) {
+
+    }
+
+    @Override
+    public void fileFinished(AuditEvent event) {
+
+    }
+
+    @Override
+    public void addError(AuditEvent event) {
```

Comment:
it is better to listen event of https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java to have exact DetailAST.
Xpath does not make sense(doe snot exist) out of TreeWalker, so it is more reliable to be listener/filter of TreeWalker.

Reply:
> so it is more reliable to be listener/filter of TreeWalker.

I didn't understand, do you suggest to replace listener with  `TreeWalkerFilter`? or there is another special type of `AuditListener` for `TreeWalker` events?

Reply:
Yes, reuse Filters to receive call on event. 

but now I see a nuance , filters is Treewlker are stored as Set - https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L72
No no guaranty to be executed the last, so it mean that our injected filter can be executed first and receive all events even user already suppressed/filtered bunch of violations. So we could print suppression queries on violations that are already suppressed by user.
As additional verification, we can do custom AuditListener in addition, and just remove all queries that were reported in Filter but not present in AuditListener. In other words CustomFilter generate xpath queries for all passed to it violations, CustomAuditListener for each violation search in query at CustomFilter and print it.

Reply:
I will try